crates/valence_protocol/build/block.rs

```diff
@@ -373,6 +398,69 @@ pub fn build() -> anyhow::Result<TokenStream> {
         })
         .collect::<TokenStream>();
 
+    let block_entity_kind_variants = block_entity_types
+        .iter()
+        .map(|block_entity| {
+            let name = ident(block_entity.name.to_pascal_case());
+            let doc = format!(
+                "The block entity type `{}` (ID {}).",
+                block_entity.name, block_entity.id
+            );
+            quote! {
+                #[doc = #doc]
+                #name,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let block_entity_kind_from_id_arms = block_entity_types
+        .iter()
+        .map(|block_entity| {
+            let id = block_entity.id;
+            let name = ident(block_entity.name.to_pascal_case());
+
+            quote! {
+                #id => Some(Self::#name),
+            }
+        })
+        .collect::<TokenStream>();
+
+    let block_entity_kind_to_id_arms = block_entity_types
+        .iter()
+        .map(|block_entity| {
+            let id = block_entity.id;
+            let name = ident(block_entity.name.to_pascal_case());
+
+            quote! {
+                Self::#name => #id,
+            }
+        })
+        .collect::<TokenStream>();
+
+    let block_entity_kind_from_ident_arms = block_entity_types
+        .iter()
+        .map(|block_entity| {
+            let name = ident(block_entity.name.to_pascal_case());
+            let ident = &block_entity.ident;
+
+            quote! {
+                #ident => Some(Self::#name),
+            }
+        })
+        .collect::<TokenStream>();
+
+    let block_entity_kind_to_ident_arms = block_entity_types
+        .iter()
+        .map(|block_entity| {
+            let name = ident(block_entity.name.to_pascal_case());
+            let ident = &block_entity.ident;
+
+            quote! {
+                Self::#name => Ident::new(#ident).unwrap(),
```

Comment:
@rj00a idea I'd like to have your opinion on:
- We create an unsafe function `Ident::new_unchecked` that takes in both the string and the index (second field) and call that here
Because right now everytime you call `BlockEntityKind::ident` it needs to parse the entire string.
Instead we parse it here during compile time. Not sure if it's worth it, just an idea I had